formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/ProtoTypes.kt

```diff
@@ -29,18 +37,48 @@ public enum class ProtoNumberType {
 @Target(AnnotationTarget.PROPERTY)
 public annotation class ProtoType(public val type: ProtoNumberType)
 
-internal typealias ProtoDesc = Pair<Int, ProtoNumberType>
+internal typealias ProtoDesc = Long
 
-internal fun extractParameters(descriptor: SerialDescriptor, index: Int, zeroBasedDefault: Boolean = false): ProtoDesc {
-    val protoId = descriptor.findAnnotation<ProtoId>(index)
-    val idx = protoId?.id ?: (if (zeroBasedDefault) index else index + 1)
-    val format = descriptor.findAnnotation<ProtoType>(index)?.type
-            ?: ProtoNumberType.DEFAULT
-    return idx to format
+@Suppress("NOTHING_TO_INLINE")
+internal inline fun ProtoDesc(protoId: Int, type: ProtoNumberType): ProtoDesc {
+    return type.signature or protoId.toLong()
 }
 
-public class ProtobufDecodingException(message: String) : SerializationException(message)
+internal inline val ProtoDesc.protoId: Int get() = (this and Int.MAX_VALUE.toLong()).toInt()
+
+internal val ProtoDesc.numberType: ProtoNumberType get() {
+    if (ProtoNumberType.DEFAULT.equalTo(this)) return ProtoNumberType.DEFAULT
```

Comment:
This probably should be more performant when expressed as bitshift with `when(mask)` instead of several if-elses